### ISSUE TITLE
Do not send a request for translation unless the user decides to

### ISSUE THREAD

**Issue by pasqoo** [sienori/simple-translate#228]
The addon's options let the user decide when a translation is shown.

The user can, for example, select some text, and a popup with a translation would show up.
The user can also choose to, instead, see a popup containing only a blue/white button, the simple-translate logo.

I thought that that would delay the request until I'd click such button.
But, analyzing the browser's console, that doesn't seem to be the case (I am using Firefox).

It's easy to imagine cases where we select content that we do not desire to be sent to the translation services (e.g. google translate).

It could be names, addresses, one-time password, etc.

I think the request should not automatically be fired in the case of the popup with the button. But it should be fired once the user clicks on the button.

What do you think?

**Comment by mrsshr** [comment]
Try uncheck 'Automatically switch to the second language'

**Comment by pbodnar** [comment]
> Try uncheck 'Do not display the button if translation is not required'

Yes, that's the (logical) way to resolve this issue, thank you!

@sienori, what do you think about **warning users** about this possible security issue when they activate this option on the settings page? Is it OFF by default, at least?

**Comment by pasqoo** [comment]
> Try uncheck 'Do not display the button if translation is not required'

In that case I would lose the button. Which is fine, to me. It would be nice not to lose it, though.

I started this discussion because I thought that the whole point of such button was to give the user the chance of postponing the translation request. This expectation is broken, as of now. I understand that I might be the only user expecting this. But there is a chance I am not alone.

> Yes, that's the (logical) way to resolve this issue, thank you!

Please take into consideration the fact that people might expect a different behavior. In fact, I have kept this extension disabled since then.

> @sienori, what do you think about **warning users** about this possible security issue when they activate this option on the settings page? Is it OFF by default, at least?

I think this behavior is present in all cases. Not just with the button. But I might be wrong. I haven't verified since then.

A red warning message would be a step forward.
The warning should clearly state WHAT triggers a translation request (meaning, WHEN users' data is sent).
Clearly the button click is not the trigger, as of now (or then? I haven't checked since the time I wrote the issue).
Please better document this "prefetch" behavior.

This is an extension that runs also on Firefox. Users of the Firefox browser have high expectations of privacy.

An example comes to my mind: some web browsers offer a feature called "page prefetch". Which, on Google Chrome, is enabled by default. But users are informed in the settings, and have a chance to disable it.
https://duckduckgo.com/?q=page+prefetch+google+chrome

I would expect the same kind of option from this extension.

**Comment by pbodnar** [comment]
@pasqoo, I actually **verified** that UNchecking 'Do not display the button if translation is not required' _does_ help and you _don't_ lose any button as you write in your comment.

Next, you can do the verification of (not) contacting Google Translate yourself: after you open Firefox Debugger (F12), go to 'XHR Breakpoints' and check 'Pause on any URL'. Then try to use the extension and see when exactly the extension makes the Ajax request to Google Translate.

Therefore I think the topic doesn't require that long discussion. Don't hesitate to correct me if you can see a different behavior of the extension though...

**Comment by pasqoo** [comment]
I now verified and it seems that the behavior is as I would expect.

I already had the **Do not display the button if translation is not required** option unchecked. I thought I had not done it, before.

I don't know what to say. I remember looking at a console (back then) and seeing requests being fired even though I was not clicking the button. Hence the effort creating this issue for it.

Sorry for the noise, and thanks for the support! @mrsshr @pbodnar

**Comment by pbodnar** [comment]
OK, thanks for the recheck! What about leaving the issue open at least as a request for giving **a clear security warning**, as discussed above?

**Comment by pasqoo** [comment]
Feel free to.

A way to do this could be:
- Move **Do not display the button if translation is not required** right below **Display translation button**
- Rename **Do not display the button if translation is not required** to **Only if required** (in order to remove the double negative)
- Keep the description, and possibly add the warning we have discussed

Personally, I would prefer to see that option gone. I find it confusing. My reasoning is: if someone chooses to see a button, it is because he wants to control when to send a request. And having that check mark enabled breaks that expectation. And you don't even notice, because the button does simply not show up.

Currently, it is enabled by default: https://github.com/sienori/simple-translate/blob/master/src/settings/defaultSettings.js#L85

Translation services are also not perfect at understanding when a translation is required. Especially for languages that have words in common, but yielding different meanings. Not all users are translating from Russian to Chinese :smile:

**Comment by maliayas** [comment]
So the problem here comes down to a naming confusion. What we want here is "make the request only after the button is clicked".

And the option "Do not display the button if translation is not required" secretly does that along with its own task. These 2 tasks here is totally unrelated as I understand:

- __Do not display the button if translation is not required__: I understand that if "from" and "to" languages are the same, hide the button.
- The proposed __Make the request only after the button is clicked__: When this is checked (which should be default due to privacy) no request is done before the user clicks the popup button; only after clicking it.

A new option "Make the request only after the button is clicked" checkbox would fix the issue IMHO.

**Comment by alurm** [comment]
Hi. Thanks for the extension.

Let me share that I copied passwords (stored in plain) while using this extension.

At some point today, I paused to think why the translation appears immediatly after clicking on the translation button.

Now I understand that it fires a request even if a button to translate is not clicked.

So I sent some of my passwords to Google in URL query parameters (as I see in the debugger window for the extension).

I expected the default behaviour to send a request only when the translate button is clicked.

I don't think the default behaviour should be to send selections to Google without a button being clicked.
